core/src/test/java/org/apache/iceberg/TestManifestListVersions.java

```diff
@@ -150,25 +163,140 @@ public void testV2Write() throws IOException {
+    assertThat(manifest.path()).isEqualTo(PATH);
+    assertThat(manifest.length()).isEqualTo(LENGTH);
+    assertThat(manifest.partitionSpecId()).isEqualTo(SPEC_ID);
+    assertThat(manifest.content()).isEqualTo(ManifestContent.DATA);
+    assertThat(manifest.sequenceNumber()).isEqualTo(SEQ_NUM);
+    assertThat(manifest.minSequenceNumber()).isEqualTo(MIN_SEQ_NUM);
+    assertThat(manifest.snapshotId()).isEqualTo(SNAPSHOT_ID);
+    assertThat(manifest.addedFilesCount()).isEqualTo(ADDED_FILES);
+    assertThat(manifest.addedRowsCount()).isEqualTo(ADDED_ROWS);
+    assertThat(manifest.existingFilesCount()).isEqualTo(EXISTING_FILES);
+    assertThat(manifest.existingRowsCount()).isEqualTo(EXISTING_ROWS);
+    assertThat(manifest.deletedFilesCount()).isEqualTo(DELETED_FILES);
+    assertThat(manifest.deletedRowsCount()).isEqualTo(DELETED_ROWS);
+    assertThat(manifest.firstRowId()).isEqualTo(WRITER_FIRST_ROW_ID);
+  }
+
+  @Test
+  public void testV3WriteMixedRowIdAssignment() throws IOException {
+    ManifestFile missingFirstRowId =
+        new GenericManifestFile(
+            PATH,
+            LENGTH,
+            SPEC_ID,
+            ManifestContent.DATA,
+            SEQ_NUM,
+            MIN_SEQ_NUM,
+            SNAPSHOT_ID,
+            PARTITION_SUMMARIES,
+            KEY_METADATA,
+            ADDED_FILES,
+            ADDED_ROWS,
+            EXISTING_FILES,
+            EXISTING_ROWS,
+            DELETED_FILES,
+            DELETED_ROWS,
+            null);
+
+    // write uses firstRowId=WRITER_FIRST_ROW_ID and ADDED_ROWS are assigned twice
+    long nextRowId = WRITER_FIRST_ROW_ID + ADDED_ROWS + ADDED_ROWS;
+    List<ManifestFile> manifests =
+        ManifestLists.read(
+            writeManifestList(3, nextRowId, missingFirstRowId, TEST_MANIFEST, missingFirstRowId));
+
+    // all v2 fields should be read correctly
+    for (ManifestFile manifest : manifests) {
+      assertThat(manifest.path()).isEqualTo(PATH);
+      assertThat(manifest.length()).isEqualTo(LENGTH);
+      assertThat(manifest.partitionSpecId()).isEqualTo(SPEC_ID);
+      assertThat(manifest.content()).isEqualTo(ManifestContent.DATA);
+      assertThat(manifest.sequenceNumber()).isEqualTo(SEQ_NUM);
+      assertThat(manifest.minSequenceNumber()).isEqualTo(MIN_SEQ_NUM);
+      assertThat(manifest.snapshotId()).isEqualTo(SNAPSHOT_ID);
+      assertThat(manifest.addedFilesCount()).isEqualTo(ADDED_FILES);
+      assertThat(manifest.addedRowsCount()).isEqualTo(ADDED_ROWS);
+      assertThat(manifest.existingFilesCount()).isEqualTo(EXISTING_FILES);
+      assertThat(manifest.existingRowsCount()).isEqualTo(EXISTING_ROWS);
+      assertThat(manifest.deletedFilesCount()).isEqualTo(DELETED_FILES);
+      assertThat(manifest.deletedRowsCount()).isEqualTo(DELETED_ROWS);
+    }
+
+    assertThat(manifests.get(0).firstRowId()).isEqualTo(WRITER_FIRST_ROW_ID);
+    assertThat(manifests.get(1).firstRowId()).isEqualTo(FIRST_ROW_ID);
```

Comment:
nit: FIRST_ROW_ID => TEST_MANIFEST_FIRST_ROW_ID

Otherwise this is a bit confusing to read

Reply:
That's not how the other constants are named :face_with_diagonal_mouth:

Reply:
Good idea. I'll update this.